sdk/src/azure/core/az_log.c

```diff
@@ -16,75 +16,51 @@
 
-void az_log_set_classifications(az_log_classification const classifications[])
-{
-  _az_log_classifications = classifications;
-}
-
-void az_log_set_callback(az_log_message_fn az_log_message_callback)
+void az_log_set_message_callback(az_log_message_fn az_log_message_callback)
 {
   _az_log_message_callback = az_log_message_callback;
 }
 
-// _az_LOG_WRITE_engine is a function private to this .c file; it contains the code to handle
-// _az_LOG_SHOULD_WRITE & _az_LOG_WRITE.
-//
-// If log_it is false, then the function returns true or false indicating whether the message
-// should be logged (without actually logging it).
-//
-// If log_it is true, then the function logs the message (if it should) and returns true or
-// false indicating whether it was logged.
-static bool _az_log_write_engine(bool log_it, az_log_classification classification, az_span message)
+void az_log_set_filter_callback(az_log_should_write_fn az_log_should_write_callback)
 {
-  // Copy the volatile fields to local variables so that they don't change within this function
-  az_log_message_fn const callback = _az_log_message_callback;
-  az_log_classification const* classifications = _az_log_classifications;
-
-  if (callback == NULL)
-  {
-    // If no one is listening, don't attempt to log.
-    return false;
-  }
-
-  if (classifications == NULL)
-  {
-    // If the user hasn't registered any classifications, then we log everything.
-    classifications
-        = &classification; // We don't need AZ_LOG_END_OF_LIST to be there, as very first comparison
-                           // is going to succeed and return from the function.
-  }
-
-  for (az_log_classification const* cls = classifications; *cls != AZ_LOG_END_OF_LIST; ++cls)
-  {
-    // If this message's classification is in the customer-provided list, we should log it.
-    if (*cls == classification)
-    {
-      if (log_it)
-      {
-        callback(classification, message);
-      }
-
-      return true;
-    }
-  }
-
-  // This message's classification is not in the customer-provided list; we should not log it.
-  return false;
+  _az_log_should_write_callback = az_log_should_write_callback;
 }
 
 // This function returns whether or not the passed-in message should be logged.
 bool _az_log_should_write(az_log_classification classification)
 {
-  return _az_log_write_engine(false, classification, AZ_SPAN_EMPTY);
+  _az_PRECONDITION(classification > 0);
+
+  // Copy the volatile fields to local variables so that they don't change within this function.
+  az_log_message_fn const message_callback = _az_log_message_callback;
+  az_log_should_write_fn const should_write_callback = _az_log_should_write_callback;
+
+  // If the user hasn't registered a should_write_callback, then we log everything, as long as a
+  // message_callback metho was provided.
+  return (should_write_callback == NULL || should_write_callback(classification))
+      && message_callback != NULL;
```

Comment:
You should capture the function pointer in a local variable first & then compare to NULL and invoke it. As written, the code suffers from a potential race condition where this thread see not NULL, another changes the pointer to NULL and then you invoke NULL causing an access violation.

You could also remove the NULL check if you guarantee that the global variable always refers to a valid function. For example, we could have a function n core that always returns true (or false).

Reply:
I think checking for NULL is better than empty callback: less cycles spent on each place that logs when user is not listening, plus, we don't pointer-jump unless there's a callback, which is also good because they may invalidate CPU cache.